Add simple end2end testing with playwright

fjahr opened the pull request.

sectore commented:
Got following error:
````bash
browserType.launch: Executable doesn't exist at /root/.cache/ms-playwright/chromium-1055/chrome-linux/chrome
````

Full output:
```bash
sudo just e2e-test
# Setup
cd docker && docker compose up -d
[+] Running 8/0
 ✔ Container ol-live-mempool-web-1   Running                                                                                                                                              0.0s 
 ✔ Container ol-live-electrs-1       Running                                                                                                                                              0.0s 
 ✔ Container ol-live-bitcoin-core-1  Running                                                                                                                                              0.0s 
 ✔ Container ol-live-ord-1           Running                                                                                                                                              0.0s 
 ✔ Container ol-live-fixtures-1      Running                                                                                                                                              0.0s 
 ✔ Container ol-live-mempool-db-1    Running                                                                                                                                              0.0s 
 ✔ Container ol-live-mempool-api-1   Running                                                                                                                                              0.0s 
 ✔ Container ol-live-nginx-proxy-1   Running                                                                                                                                              0.0s 
just watch > /dev/null 2>&1 & echo $! > pidfile # record pid to kill process later
sleep 5 # Wait for just watch finish starting
# test
node test/end2end/runner.ts
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

browserType.launch: Executable doesn't exist at /root/.cache/ms-playwright/chromium-1055/chrome-linux/chrome
╔═════════════════════════════════════════════════════════════════════════╗
║ Looks like Playwright Test or Playwright was just installed or updated. ║
║ Please run the following command to download new browsers:              ║
║                                                                         ║
║     npx playwright install                                              ║
║                                                                         ║
║ <3 Playwright Team                                                      ║
╚═════════════════════════════════════════════════════════════════════════╝
    at ordilabs/repos/live/test/end2end/runner.ts:4:34
    at Object.<anonymous> (ordilabs/repos/live/test/end2end/runner.ts:22:3) {
  name: 'Error'
}

Node.js v18.15.0
error: Recipe `e2e-test` failed on line 173 with exit code 1
```

Notes:
- `npx playwright install` did not help
- I have to run `just e2e-test` with `sudo`. In other case I got a permission issue: 
   ```bash 
   just watch > /dev/null 2>&1 & echo $! > pidfile # record pid to kill process later
   sh: 1: cannot create pidfile: Permission denied
   ```
 - Test on `Ubuntu 22.04.2 LTS`

sectore commented:
Update `#test` section with following works to me (Note: `node` vs. `node-ts`): 
````bash
    # test
    npx playwright install
    npx ts-node test/end2end/runner.ts
```
works now. 

In case use `sudo` can't find `node`, `nvm` and/or `npx` aliases are needed

```bash
sudo ln -s "$NVM_DIR/versions/node/$(nvm version)/bin/node" "/usr/local/bin/node"
sudo ln -s "$NVM_DIR/versions/node/$(nvm version)/bin/nvm" "/usr/local/bin/nvm"
sudo ln -s "$NVM_DIR/versions/node/$(nvm version)/bin/npx" "/usr/local/bin/npx"
```
^ https://github.com/nvm-sh/nvm/issues/43#issuecomment-818700757

sectore commented on test/end2end/runner.ts:
```diff
@@ -0,0 +1,22 @@
+  const context = await browser.newContext();
+  const page = await context.newPage();
+
+  await page.goto('http://live-ol.local');
```
Port `3000` is needed to me to fix https://github.com/ordilabs/live/pull/86#issuecomment-1514941652 

```bash
# test
npx playwright install
npx ts-node test/end2end/runner.ts
Test passed
```

sectore commented:
Only (last) issue now:
 
```bash
# Teardown
kill $(cat pidfile) && rm pidfile # kill watch process
sh: 1: kill: No such process

error: Recipe `e2e-test` failed on line 173 with exit code 1
```

fjahr commented:
@sectore Did you try running `just install` first? I added `playwright` in the npm dependencies and that was enough on my setup. I have added a note in the readme on typescript as an additional dependency.

fjahr commented:
> Only (last) issue now:
> 
> ```shell
> # Teardown
> kill $(cat pidfile) && rm pidfile # kill watch process
> sh: 1: kill: No such process
> 
> error: Recipe `e2e-test` failed on line 173 with exit code 1
> ```

Looks like a second order effect when the watch process wasn't actually started. Did the test pass on this run?

fjahr commented:
> ```shell
> cat /etc/hosts
> ```

I have the same, for me it wasn't needed. Can you suggest changes that make it work in your environment?

```
127.0.0.1 mempool-ol.local
::1 mempool-ol.local
127.0.0.1 ord-ol.local
::1 ord-ol.local
127.0.0.1 live-ol.local
::1 live-ol.local
```

fjahr commented:
So is this in general the setup that we want conceptually and I can remove the draft mode? @FelixWeis

FelixWeis commented:
Aproach ACK. few ideas

- [ ] docker compose should a testing namespace to not tear down less ephemeral dev fixtures
- [ ] instance `just watch` on a different port than :3000 using LEPTOS_SITE_ADDR  env
- [ ] these setting probably should all just be in a testing.env that `just` loads

sectore commented:
@fjahr Trying to answer your questions here:
> Did you try running just install first? 

Yes

> Can you suggest changes that make it work in your environment?

I've just added `http://live-ol.local:3000` to include port `3000` in line `8` of `test/end2end/runner.ts`

> Looks like a second order effect when the watch process wasn't actually started. Did the test pass on this run?

Yes, tests did pass.

fjahr commented:
@sectore addressed your comments as far as I could without breaking my own environment, please give it another try.

But I guess @FelixWeis is right, this should be dockerized. I haven't worked with docker for a while so this would take me some time. How about you @sectore ? Are you familiar with Docker and want to take it over from here?

sectore commented on README.md:
```diff
@@ -10,6 +10,7 @@ We use Docker to containerize all micro-services and you `just` need to run a fe
 
 - [Rust ^1.68](https://rustup.rs/)
 - [nodejs ^18.15](https://nodejs.org/en)
+- [Typescript](https://www.typescriptlang.org/)
```
To avoid mismatch of users global installed TypeScript version I would recommend to put it into `devDependencies` of `package.json`.

sectore commented:
> How about you @sectore ? Are you familiar with Docker and want to take it over from here?

Sorry, I'm not that deep into Docker as well.

FelixWeis commented:
Thanks until here! Assigning this to myself now to explore the dockerization.
